diff --git a/dtoa.c b/dtoa.c
index ac1be8d..9732a77 100644
--- a/dtoa.c
+++ b/dtoa.c
@@ -31 +30,0 @@
-#include <sys/time.h>
diff --git a/quickjs.c b/quickjs.c
index fa5a7a9..b8de5f3 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -30,0 +31,3 @@
+#if defined(_WIN32)
+#include <windows.h>
+#else
@@ -31,0 +35 @@
+#endif
@@ -35 +39,3 @@
-#if defined(__APPLE__)
+#if defined(_WIN32)
+#include <malloc.h>
+#elif defined(__APPLE__)
@@ -71 +77 @@
-#if !defined(__EMSCRIPTEN__)
+#if !defined(__EMSCRIPTEN__) && !defined(_WIN32)
@@ -625 +631 @@ typedef struct JSClosureVar {
-    JSClosureTypeEnum closure_type : 3;
+    uint8_t closure_type : 3; /* see JSClosureTypeEnum */
@@ -47146 +47152 @@ static uint64_t xorshift64star(uint64_t *pstate)
-static void js_random_init(JSContext *ctx)
+static int64_t js_get_utc_time_us(void)
@@ -47147,0 +47154,10 @@ static void js_random_init(JSContext *ctx)
+#if defined(_WIN32)
+    FILETIME file_time;
+    ULARGE_INTEGER file_time_value;
+
+    GetSystemTimeAsFileTime(&file_time);
+    file_time_value.LowPart = file_time.dwLowDateTime;
+    file_time_value.HighPart = file_time.dwHighDateTime;
+    return (int64_t)((file_time_value.QuadPart -
+                      UINT64_C(116444736000000000)) / 10);
+#else
@@ -47148,0 +47165 @@ static void js_random_init(JSContext *ctx)
+
@@ -47150 +47167,7 @@ static void js_random_init(JSContext *ctx)
-    ctx->random_state = ((int64_t)tv.tv_sec * 1000000) + tv.tv_usec;
+    return ((int64_t)tv.tv_sec * 1000000) + tv.tv_usec;
+#endif
+}
+
+static void js_random_init(JSContext *ctx)
+{
+    ctx->random_state = js_get_utc_time_us();
@@ -55171,3 +55194 @@ static int64_t date_now(void) {
-    struct timeval tv;
-    gettimeofday(&tv, NULL);
-    return (int64_t)tv.tv_sec * 1000 + (tv.tv_usec / 1000);
+    return js_get_utc_time_us() / 1000;
